stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/avatar/AvatarView.kt

```diff
@@ -0,0 +1,131 @@
+package io.getstream.chat.android.ui.avatar
+
+import android.content.Context
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.Path
+import android.graphics.RectF
+import android.util.AttributeSet
+import androidx.appcompat.widget.AppCompatImageView
+import io.getstream.chat.android.client.models.Channel
+import io.getstream.chat.android.client.models.User
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+public class AvatarView @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : AppCompatImageView(context, attrs, defStyleAttr) {
+    private val bitmapFactory = AvatarBitmapFactory(context)
+    private val clipPath = Path()
+    private val clipRect = RectF()
+    private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
+        style = Paint.Style.STROKE
+    }
+    private val onlineIndicatorOutlinePaint = Paint().apply {
+        style = Paint.Style.FILL
+        color = Color.WHITE
+    }
+    private val onlineIndicatorPaint = Paint().apply {
+        style = Paint.Style.FILL
+        color = Color.GREEN
+    }
+
+    private lateinit var avatarStyle: AvatarStyle
+    private var onlineIndicatorVisible: Boolean = false
+
+    init {
+        setStyle(AvatarStyle(context, attrs))
+    }
+
+    public fun setChannelData(channel: Channel, user: List<User>) {
+        configureImageDrawable {
+            AvatarDrawable(bitmapFactory.createChannelBitmaps(channel, user, avatarStyle))
+        }
+    }
+
+    public fun setUserData(user: User) {
+        configureImageDrawable {
+            AvatarDrawable(listOfNotNull(bitmapFactory.createUserBitmap(user, avatarStyle)))
+        }
+    }
+
+    public fun toggleOnlineIndicatorVisibility(visible: Boolean) {
+        this.onlineIndicatorVisible = visible
+        invalidate()
+    }
+
+    public fun setStyle(avatarStyle: AvatarStyle) {
+        this.avatarStyle = avatarStyle
+        borderPaint.color = avatarStyle.avatarBorderColor
+        borderPaint.strokeWidth = avatarStyle.avatarBorderWidth.toFloat()
+    }
+
+    private fun configureImageDrawable(generateAvatarDrawable: suspend () -> AvatarDrawable) {
+        GlobalScope.launch(Dispatchers.Main) {
+            layoutParams?.apply {
+                width = avatarStyle.avatarWidth
+                height = avatarStyle.avatarHeight
+            }?.let(::setLayoutParams)
+            setImageDrawable(generateAvatarDrawable())
```

Comment:
- The first couple lines here are equivalent to an `updateLayoutParams` call from `Dependencies.androidxCoreKtx`
- Why do we need to update layout params before setting the new drawable?
- Is it okay that there might be some amount of time between `setLayoutParams` and `setImageDrawable` being called? `generateAvatarDrawable` is suspending here. If we want to call the two setters at the same time (for the same frame), we should create the Drawable first, and then call them.

Reply:
Done